docs/src/tutorial.md

```diff
@@ -0,0 +1,301 @@
+```@meta
+DocTestFilters = [
+    r"┌ @ ((REPL\[\d+\])|(none)).+$",
+    r"│*└─+\s*$",
+]
+DocTestSetup = quote
+    using JET
+end
+```
+
+# JET tutorial
+JET leverages the Julia compiler's inference to check user code for type instability and type errors.
+This tutorial will demonstrate how to use JET effectively.
+It presupposes the reader has working knowledge of Julia, and understands Julian concepts such as "dynamic dispatch" and "type stability" - see the [Julia documentation](https://docs.julialang.org/en/v1/) for explanation of these concepts.
+
+Because JET relies on the compiler's type inference, it is not able to effectively analyze type unstable code.
+Making your code type stable is a prerequisite for effectively using JET's type error analysis.
+Therefore, we will begin by showing how to use JET to fix type instabilities.
+
+## Detecting type instability with `@report_opt`
+JET exports a function `report_opt` and the related macro `@report_opt`.
+It works similar to the function/macro pair `(@)code_warntype` from Base - except that it automatically analyses all the way down the function chain, and that it only displays any issues found.
+
+For example, suppose we have the function:
+```jldoctests opt1
+julia> add_one_first(x) = first(x) + 1;
+```
+
+Any type instabilities of a given function call can be analysed thus:
+```jldoctests opt1
+julia> @report_opt add_one_first([1])
+No errors detected
+
+
+julia> @report_opt add_one_first(Any[1])
+═════ 1 possible error found ═════
+┌ @ REPL[8]:1 %1 + 1
+│ runtime dispatch detected: (%1::Any + 1)::Any
+└─────────────
+```
```

Comment:
```suggestion
\`\`\`@repl opt1
@report_opt add_one_first([1])

@report_opt add_one_first(Any[1])
\`\`\`
```

We can use [`@repl`](https://documenter.juliadocs.org/stable/man/syntax/#@repl-block) block to automatically generate outputs (I'm not a big fan of `jldoctests`...).